### PR TITLE
Introduce int casts

### LINKED ISSUE #554: int casts
```
$a = (int)0;
```

Or whatever int type.

### PR DESCRIPTION
**Pull request by fbs**
Currently all integers are 64 bits internally. This causes issues when
dealing with smaller (bitsize) values returned by the kernel. An example
is:

```
kr:sock_recvmsg /retval < 0/{
  @[comm, retval]++;
}
```

`sock_recvmsg` returns a 32 bits int which, depending on the CPU
architecture, gets zero extended to 64 bits. Because of the zero
extension predicate will always be false.
With int casts the user can tell bpftrace that it should cast the value
before doing the comparison.

My idea is to keep all integer 64bit internally, but keep track of
whether they're signed by extending `SizedType`.  The sign information
can then be used in the all the CreateIntCasts to correctly extend the
value to 64 bits. All that is required for a cast then is a single
truncate instruction. E.g. for a simple predicate like
`'kr:sock_recvmsg /(int32) retval < 0/ {}` the generated code will be

```
  %3 = load i64, i64* %retval
  %cast = trunc i64 %3 to i32
  %5 = sext i32 %cast to i64
  %6 = icmp slt i64 %5, 0
  %7 = zext i1 %6 to i64
```

Opening this early to get feedback on the approach I'm taking here. 

This fixes #554

### CODE REVIEW

**Comment by brendangregg** [comment]
Tested, works for me. And thanks, I need this right now. :-)

I think the "int32" style is ok. More desciptive than i32, and less clutter than int32_t.

Just so I understand what's going on here: the kprobe/kretprobe arguments & retval get zero extended to 64-bit, because, well, we're just reading the 64-bit registers. 

Thinking ahead: in the future we'll have debuginfo/BTF, and can know the return type. Here's what I propose (this is way early, I know):

- retval: stays as a uint64, and is documented as such.
- retval_t (or another name): is a typed version of retval, so in this case it would be int32. If you didn't have debuginfo/BTF, it would throw an error "you need BTF to use retval_t".

With this plan, the following one-liner makes sense both today and in the future, because retval is always a uint64 and always needs casting:

```kr:sock_recvmsg /(int32)retval < 0/ { @[comm, retval]++; }```

(Which means I can write docs using this now, and it won't change. :)

**Comment by fbs** [comment]
> Just so I understand what's going on here: the kprobe/kretprobe arguments & retval get zero extended to 64-bit, because, well, we're just reading the 64-bit registers.

What happens on x86-64 is that the compiler emits  32 bits instructions when returning an 32 bit value. E.g. for:

```
int fnint() {
  return 0xffffffffffffffff;
}

long int fnlong() {
  return 0xffffffffffffffff;
}

---

0000000100000f00 <_fnint>:
...
   100000f07:	b8 ff ff ff ff       	mov    $0xffffffff,%eax
   100000f0c:	5d                   	pop    %rbp
   100000f0d:	c3                   	retq

0000000100000f10 <_fnlong>:
...
   100000f18:	48 c7 c0 ff ff ff ff 	mov    $0xffffffffffffffff,%rax
   100000f1f:	5d                   	pop    %rbp
   100000f20:	c3                   	retq
```

And the CPU  fills the [upper 32 bits with zeroes when using 32 bit](https://stackoverflow.com/questions/11177137/why-do-x86-64-instructions-on-32-bit-registers-zero-the-upper-part-of-the-full-6) . eBPF is 64 bit internally and just copies in the register, it doesn't know it was 32 bits originally.

So on other architectures it might not be needed, I haven't looked into that. 

>` kr:sock_recvmsg /(int32)retval < 0/ { @[comm, retval]++; }`

That should be `kr:sock_recvmsg /(int32)retval < 0/ { @[comm, (int32)retval]++; }` as the type of retval doesn't change. You need to cast it for every use. Or  ` >= 0` if it was a typo :)

**Comment by brendangregg** [comment]
Would be good to get this finished; just ran into the lack of casts again.

**Comment by fbs** [comment]
Yes. Been a bit busy but have some time for it the coming week.

Could use some input, both on the approach taken and on the cases to test.  With the latests commit everything I could think of seems to work. When there is agreement on the approach taken I will look into useful semantic and runtime tests to add and clean it up where needed.

So far this is what I've tested and they all seem to work fine. Are there more edge cases I should check?

## Socket.py

Run in background

```
import socket
import time

host = ('127.0.0.1', 12345)

snd = socket.socket()
snd.bind(host)
snd.listen(1)

rcv = socket.socket()
rcv.connect(host)

while True:
   try:
       rcv.recv(1, socket.MSG_DONTWAIT)
   except:
       pass
   time.sleep(1)
```

## Int output should be sorted by value

```
sudo ./src/bpftrace -e 'BEGIN { @[10]=-10; @[5]=0; @[0]=10; @[-10]=100; exit() }'

@[10]: -10
@[5]: 0
@[0]: 10
@[-10]: 100
```

## int casting should work

```
sudo ./src/bpftrace -e 'kr:sock_recvmsg /(int32) retval < 0/ { @[comm] = count() }'

@[ntpd]: 9
@[socket]: 58
```

## Negative map keys

```
sudo ./src/bpftrace -e 'BEGIN { @[10]=-10; @[5]=0; @[0]=10; @[-10]=100; exit() }'

@[10]: -10
@[5]: 0
@[0]: 10
@[-10]: 100
```

```
sudo ./src/bpftrace -e 'kr:sock_recvmsg /(int32) retval < 0/ { @[(int32) retval, comm] = count() } i:s:5 { exit() }'

@[-11, socket]: 5
```

## Negative map values

```
sudo ./src/bpftrace -e 'kr:sock_recvmsg { @[comm]=(int32)retval; print(@); }'
@[python]: -11
@[dbus-daemon]: 0
@[systemd]: 4
```


## Casting should work

```
sudo ./src/bpftrace -e 'kr:sock_recvmsg /(int32) retval < 0/ { @[(int32) retval, comm] = count() } i:s:5 { exit() }'

@[-11, socket]: 5
```

## Builtins should work

### avg & stats

```
sudo ./src/bpftrace -e 'kr:sock_recvmsg{ @a=avg((int32)retval); @s=stats((int32)retval); }'

@a: 24
@s: count 13, average 24, total 316

```

### sum, min & max

```
sudo ./src/bpftrace -e 'kr:sock_recvmsg /(int32) retval < 0/ { @=sum((int32)retval); print(@); }'
Attaching 1 probe...
@: -11
@: -22
```

### hist&lhist

Don't support negative yet but should not show a massive number

```
sudo ./src/bpftrace -e 'kr:sock_recvmsg { @=hist((int32)retval) }'
Attaching 1 probe...
^C

@:
(..., 0)               3 |@@@@@@@@@@@@@@@@@@@@@@                              |
[0]                    0 |                                                    |
[1]                    0 |                                                    |
[32, 64)               7 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
```

Additionally all current tests should pass.

File: src/ast/semantic_analyser.cpp [hunk 1]
```diff
@@ -850,10 +851,28 @@ void SemanticAnalyser::visit(FieldAccess &acc)
   }
 }
 
-void SemanticAnalyser::visit(Cast &cast)
-{
+void SemanticAnalyser::visit(Cast &cast) {
```
Comment (mmarchini) [inline src/ast/semantic_analyser.cpp]: Please don't change the styling unless you're already changing the line and the new styling matches the majority of the code base (neither is the case here), otherwise these changes make reviewing the code harder as now there are unrelated lines changed in the file.

(eventually we'll have `clang-format` to enforce a consistent format across our files)

File: src/ast/semantic_analyser.cpp [hunk 2]
```diff
@@ -862,18 +881,14 @@ void SemanticAnalyser::visit(Cast &cast)
   int cast_size;
   if (cast.is_pointer) {
     cast_size = sizeof(uintptr_t);
-  }
-  else {
+  } else {
```
Comment (mmarchini) [inline src/ast/semantic_analyser.cpp]: Ditto about code style changes.

File: src/ast/semantic_analyser.cpp [hunk 3]
```diff
@@ -862,18 +881,14 @@ void SemanticAnalyser::visit(Cast &cast)
...
   }
   cast.type = SizedType(Type::cast, cast_size, cast.cast_type);
   cast.type.is_pointer = cast.is_pointer;
 }
 
-void SemanticAnalyser::visit(ExprStatement &expr)
-{
-  expr.expr->accept(*this);
-}
+void SemanticAnalyser::visit(ExprStatement &expr) { expr.expr->accept(*this); }
```
Comment (mmarchini) [inline src/ast/semantic_analyser.cpp]: Ditto about code style changes.

File: src/bpftrace.cpp [hunk 4]
```diff
@@ -1024,12 +1024,27 @@ int BPFtrace::print_map(IMap &map, uint32_t top, uint32_t div)
     old_key = key;
   }
 
-  if (map.type_.type == Type::count || map.type_.type == Type::sum || map.type_.type == Type::integer)
+  if (map.type_.type == Type::count || map.type_.type == Type::sum ||
+      map.type_.type == Type::integer)
```
Comment (mmarchini) [inline src/bpftrace.cpp]: Ditto about code style changes. (I thought something actually changed in this line, but it's the same if)

File: src/types.cpp [hunk 5]
```diff
@@ -23,12 +23,18 @@ std::ostream &operator<<(std::ostream &os, const SizedType &type)
   }
   if (type.is_pointer)
     os << "*";
+  if (type.type == Type::integer && (type.size != 8 || type.is_signed)) {
```
Comment (mmarchini) [inline src/types.cpp]: I think we can use a consistent formatting for all ints, no need to have an exception for signed 64-bit ints.

```suggestion
  if (type.type == Type::integer) {
```

File: src/types.cpp [hunk 6]
```diff
@@ -23,12 +23,18 @@ std::ostream &operator<<(std::ostream &os, const SizedType &type)
...
+    os << "(";
+    if (type.is_signed)
```
Comment (mmarchini) [inline src/types.cpp]: Shouldn't it be

```suggestion
    if (!type.is_signed)
```

?

File: src/bpftrace.h [hunk 7]
```diff
@@ -164,7 +164,8 @@ class BPFtrace
   int print_map_stats(IMap &map);
   int print_hist(const std::vector<uint64_t> &values, uint32_t div) const;
   int print_lhist(const std::vector<uint64_t> &values, int min, int max, int step) const;
-  static uint64_t reduce_value(const std::vector<uint8_t> &value, int ncpus);
+  template <typename T> static T reduce_value(const std::vector<uint8_t> &value, int ncpus);
+  /* static uint64_t reduce_value(const std::vector<uint8_t> &value, int ncpus); */
```
Comment (mmarchini) [inline src/bpftrace.h]: Remove comment.

**Comment by fbs** [comment]
@mmarchini 

Thanks for reviewing :)

> IIUC you want #834 to land first, right? I'll review that next.

Yes I think thats better as this one is mixing two things, adding signed ints is a big enough change on its own. I'm working on #834 atm, adding tests. I will update it tonight (in a few hours)

**Comment by fbs** [comment]
I've rebased this on #834 simplified this by quite a bit, see https://github.com/iovisor/bpftrace/pull/772/commits/9bd54b09b6c20fdae77c655f81d647ab16ca149f

The way casts have been implemented has stayed the same. The biggest difference is the handling of map keys. I've chosen to insert a cast to 64bit into the AST to make sure everything behaves correctly. The alternative is to allow for smaller integer map keys smaller than 64 bits but requires modifications in more places with no clear benefit.

It could probably use a few more tests but I couldn't come up with anything. 

All existing tools seems to work just fine.
